src/core/tensor/tensor_math_cuda.h

```diff
@@ -20,260 +20,428 @@
+  float* tPtr = static_cast<float*>(t.block()->mutable_data());
+  int dimIn[in->shape().size()+1]; int stridesIn[in->strides().size()+1];
+  int dimT[t.shape().size()+1]; int stridesT[t.strides().size()+1];
+  in->shape_cuda(dimIn,8);
+  in->strides_cuda(stridesIn,8);
+  t.shape_cuda(dimT,8);
+  t.strides_cuda(stridesT,8);
+
+  //reduce_desc
+  cudnnReduceTensorDescriptor_t reduce_desc;
+  cudnnReduceTensorOp_t reduce_op = CUDNN_REDUCE_TENSOR_ADD;
+  cudnnDataType_t cudnn_dtype = CUDNN_DATA_FLOAT;
+  cudnnNanPropagation_t cudnn_propagation = CUDNN_NOT_PROPAGATE_NAN;
+  cudnnReduceTensorIndices_t cudnn_indices = CUDNN_REDUCE_TENSOR_NO_INDICES;
+  cudnnIndicesType_t cudnn_indices_type = CUDNN_32BIT_INDICES;
+  cudnnCreateReduceTensorDescriptor(&reduce_desc);
+  cudnnSetReduceTensorDescriptor(reduce_desc, reduce_op, cudnn_dtype,
+                                 cudnn_propagation, cudnn_indices, cudnn_indices_type);
+
+  //cudaMalloc
+  //size_t indices_bytes = 0; size_t workspace_bytes = 0;//need to change these variables, currently allocating all free memory
+  //void* d_indices{nullptr}; void* d_workspace{nullptr};
+  //cudaMalloc(&d_indices, indices_bytes); cudaMalloc(&d_workspace, workspace_bytes);
+
+  //instantiate new tensor to use new blocks as memory instead of cudaMalloc
+  //create 2 tensors of same size as input tensor
+  Tensor indices(in->shape(), in->data_type());
+  Tensor workspace(in->shape(), in->data_type());
+  size_t indices_bytes = indices.block()->size();
+  size_t workspace_bytes = workspace.block()->size();
+  void* indicesPtr = indices.block()->mutable_data();
+  void* workspacePtr = workspace.block()->mutable_data();
+
+  float alpha = 1.0, beta = 0.0;
+  cudnnTensorDescriptor_t in_desc, t_desc;
+  cudnnCreateTensorDescriptor(&in_desc);
+  cudnnCreateTensorDescriptor(&t_desc);
+  cudnnSetTensorNdDescriptor(in_desc, cudnn_dtype, in->nDim(), dimIn, stridesIn);
+  cudnnSetTensorNdDescriptor(t_desc, cudnn_dtype, t.nDim(), dimT, stridesT);
+  cudnnReduceTensor(ctx->cudnn_handle, reduce_desc,
+                    indicesPtr, indices_bytes, workspacePtr, workspace_bytes,
+                    &alpha, in_desc, inPtr, &beta,  t_desc, tPtr);
+
+  *out = tPtr[0]; 
+  cudnnDestroyTensorDescriptor(in_desc);
+  cudnnDestroyTensorDescriptor(t_desc);
+  //cudaFree(d_indices);
+  //cudaFree(d_workspace);
 }
 
 /// Element-wise operation, out[i]=tanh([in[i])
 template <>
-void Tanh<float, lang::Cuda>(const size_t num, const Block* in, Block* out,
+void Tanh<float, lang::Cuda>(const Tensor* in, Tensor* out,
                              Context* ctx) {
-  const float* inPtr = static_cast<const float*>(in->data());
-  float* outPtr = static_cast<float*>(out->mutable_data());
+  const float* inPtr = static_cast<const float*>(in->block()->data());
+  float* outPtr = static_cast<float*>(out->block()->mutable_data());
+  const size_t num = in->Size();
   cuda::tanh(num, inPtr, outPtr, ctx->stream);
```

Comment:
for tanh, sigmoid, exp, pow, etc , if in and out have the same stride, then call existing cuda code. otherwise, report error.